### PR TITLE
Launcher process redeploy

### PR DESCRIPTION
**Pull request by cescoffier**
This PR adds the _redeploy_ feature to the `run` command.

It's a process-based redeploy: the application is started in background and restarted on file change. The set of watched file is configured in the command line (`--redeploy=...`).

### CODE REVIEW

File: src/main/java/io/vertx/core/impl/launcher/commands/Watcher.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+  private final static Logger LOGGER = LoggerFactory.getLogger(Watcher.class);
+
+  private final List<String> includes;
+  private final File root;
+  private final Handler<Handler<Void>> deploy;
+  private final Handler<Handler<Void>> undeploy;
+  private final String cmd;
+
+  private WatchService service;
```
Comment (purplefox) [inline src/main/java/io/vertx/core/impl/launcher/commands/Watcher.java]: Beware of WatchService - we used it in the old Redeployer originally but it didn't work properly on all OSes so I removed it and wrote our own watch code.

Maybe they've fixed those bugs - but I'm not sure...

File: src/test/java/io/vertx/core/impl/launcher/commands/WatcherTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      if (files != null) {
+        for (File f : files) {
+          ret = ret && deleteRecursive(f);
+        }
+      }
+    }
+    return ret && path.delete();
+  }
+
+}
```
Comment (purplefox) [inline src/test/java/io/vertx/core/impl/launcher/commands/WatcherTest.java]: Also test addition and deletion of directory? In particular:

Create new directory
Create new file in directory.
Amend file in directory.

**Comment by purplefox** [comment]
Looks good. Just a few comments.

I'm a bit concerned about Watcher though - this didn't work well for us in the past...

File: src/main/java/io/vertx/core/impl/launcher/commands/Watcher.java (hunk 1 above)
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/launcher/commands/Watcher.java]: From my test on MacOSX, Linux and Windows it works. The polling period is different and the events are not ordered the same way, but it is ok in our case.

**Comment by cescoffier** [comment]
If we get issue because of the Watcher, we will replace it. I know that they had couple of bugs in their native integration, but they fixed them. 

However, I'm also concerns by non standard file system, or on non Oracle / OpenJDK JVM. We may want to have a home made file alteration monitor alternative in the future, with a toggle.

**Comment by cescoffier** [comment]
I've extended the tests.

**Pull request merged**